--- llvm/lib/IR/DebugInfo.cpp	2017-06-16 12:35:00.129385098 -0400
+++ DebugInfo_new.cpp	2017-06-16 12:31:59.504971600 -0400
@@ -891,7 +891,7 @@
       for (unsigned i = 0, e = RetainedTypes.getNumElements(); i != e; ++i)
         processType(DIType(RetainedTypes.getElement(i)));
       // FIXME: We really shouldn't be bailing out after visiting just one CU
-      return;
+      //return;
     }
   }
 }
