diff --git a/queens_GPU_CPU_single_node.chpl b/queens_GPU_CPU_single_node.chpl
index 088cd32..6406d3b 100644
--- a/queens_GPU_CPU_single_node.chpl
+++ b/queens_GPU_CPU_single_node.chpl
@@ -62,7 +62,7 @@ proc main(){
 	 			when "serial"{
 	 				
  					writeln("--- N-Queens serial search --- \n\n");
- 					queens_call_serial_search(size, mode, prepro);
+ 					queens_call_serial_search(size, mode);
 	 			}
 			
 	 			when "mcore"{
