--- a/modules/queens_call_serial_search.chpl
+++ b/modules/queens_call_serial_search.chpl
@@ -7,6 +7,7 @@ module queens_call_serial_search{
     use queens_constants;
     use queens_aux;
 
+  config const printElapsedTime = false;
 
 	proc queens_call_serial_search(const size: uint(16), const mode: string = "serial", const prepro: bool = false){
 
@@ -22,7 +23,9 @@ module queens_call_serial_search{
 		writeln("\nSerial Bitset Queens for size ", size);
 		writeln("\tNumber of solutions: ", metrics[0]);
 		writeln("\tTree size: ", metrics[1]);
-		writeln("Elapsed time: ", timer.elapsed()*1000, " ms"); // Print elapsed time
+    if printElapsedTime {
+		  writeln("Elapsed time: ", timer.elapsed()*1000, " ms"); // Print elapsed time
+    }
 		writeln("Performance: ", metrics[1]/timer.elapsed(), " n/s\n\n"); // Print elapsed time
 		writeln("(performance in subproblems evaluation per sec.): "); // Print elapsed time
         timer.clear(); // Clear timer for parallel loop
