--- 12-no-capture-in-begin.good.o	2013-10-22 18:46:19.000198000 -0700
+++ 12-no-capture-in-begin.good.o	2013-10-22 18:48:24.000795000 -0700
@@ -88,11 +88,6 @@
 rec2 (yy01 = 0, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 0, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 0)
 cls nil
 unn ()
-dom1 {1..0}
-dom2 {1..0, 1..0}
-arr1 0 0
-arr2 0 0
-0 0
 after begin
 b0 true
 b8 true
@@ -119,11 +114,11 @@
 rec2 (yy01 = 30001, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 30007, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 30020)
 cls {zzz = 44444444}
 unn (ufield222 = 339933)
-dom1 {7770..7771}
-dom2 {110001..110002, 330004..330005}
-arr1 11011 11011
-arr2 33033 33033
-33033 33033
+dom1 {1..0}
+dom2 {1..0, 1..0}
+arr1 0 0
+arr2 0 0
+0 0
 === in a begin ===
 in begin
 b0 false
@@ -151,11 +146,6 @@
 rec2 (yy01 = 0, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 0, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 0)
 cls nil
 unn ()
-dom1 {1..0}
-dom2 {1..0, 1..0}
-arr1 0 0
-arr2 0 0
-0 0
 after begin
 b0 true
 b8 true
@@ -182,11 +172,11 @@
 rec2 (yy01 = 30001, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 30007, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 30020)
 cls {zzz = 44444444}
 unn (ufield222 = 339933)
-dom1 {7770..7771}
-dom2 {110001..110002, 330004..330005}
-arr1 11011 11011
-arr2 33033 33033
-33033 33033
+dom1 {1..0}
+dom2 {1..0, 1..0}
+arr1 0 0
+arr2 0 0
+0 0
 === in a cobegin ===
 in begin
 b0 false
@@ -214,11 +204,6 @@
 rec2 (yy01 = 0, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 0, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 0)
 cls nil
 unn ()
-dom1 {1..0}
-dom2 {1..0, 1..0}
-arr1 0 0
-arr2 0 0
-0 0
 after begin
 b0 true
 b8 true
@@ -245,11 +230,11 @@
 rec2 (yy01 = 30001, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 30007, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 30020)
 cls {zzz = 44444444}
 unn (ufield222 = 339933)
-dom1 {7770..7771}
-dom2 {110001..110002, 330004..330005}
-arr1 11011 11011
-arr2 33033 33033
-33033 33033
+dom1 {1..0}
+dom2 {1..0, 1..0}
+arr1 0 0
+arr2 0 0
+0 0
 === in a coforall ===
 in begin
 b0 false
@@ -277,11 +262,6 @@
 rec2 (yy01 = 0, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 0, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 0)
 cls nil
 unn ()
-dom1 {1..0}
-dom2 {1..0, 1..0}
-arr1 0 0
-arr2 0 0
-0 0
 after begin
 b0 true
 b8 true
@@ -308,9 +288,9 @@
 rec2 (yy01 = 30001, yy02 = 0, yy03 = 0, yy04 = 0, yy05 = 0, yy06 = 0, yy07 = 30007, yy08 = 0, yy09 = 0, yy10 = 0, yy11 = 0, yy12 = 0, yy13 = 0, yy14 = 0, yy15 = 0, yy16 = 0, yy17 = 0, yy18 = 0, yy19 = 0, yy20 = 30020)
 cls {zzz = 44444444}
 unn (ufield222 = 339933)
-dom1 {7770..7771}
-dom2 {110001..110002, 330004..330005}
-arr1 11011 11011
-arr2 33033 33033
-33033 33033
+dom1 {1..0}
+dom2 {1..0, 1..0}
+arr1 0 0
+arr2 0 0
+0 0
 === done ===
